common/setups/hybrid/nn_system.py

```diff
@@ -0,0 +1,167 @@
+        returnn_python_exe: str = gs.RETURNN_PYTHON_EXE,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root
+        self.returnn_python_home = returnn_python_home
+        self.return_python_exe = returnn_python_exe
+
+        self.nn_args = None
+
+        self.devtrain_corpora = []
+        self.cv_corpora = []
+
+        self.base_returnn_args = None
+        self.returnn_args = []
+
+    # -------------------- Setup --------------------
+    def init_system(
+        self,
+        hybrid_init_args: HybridInitArgs,
+        nn_args: NnArgs,
+        train_data: Dict[str, HybridDataInput],
+        devtrain_data: Dict[str, HybridDataInput],
+        cv_data: Dict[str, HybridDataInput],
+        dev_data: Dict[str, HybridDataInput],
+        test_data: Dict[str, HybridDataInput],
+    ):
+        self.hybrid_init_args = hybrid_init_args
+        self.nn_args = nn_args
+
+        self._init_am(**self.hybrid_init_args.am_args)
+
+        self._assert_corpus_name_unique(
+            train_data, devtrain_data, cv_data, dev_data, test_data
+        )
+
+        for name, v in sorted(train_data.items()):
+            self.add_corpus(name, data=v, add_lm=False)
+            self.train_corpora.append(name)
+            assert self.concurrent[name] == 1, "concurrency for train data should be 1"
+
+        for name, v in sorted(devtrain_data.items()):
+            self.add_corpus(name, data=v, add_lm=False)
+            self.devtrain_corpora.append(name)
+            assert (
+                self.concurrent[name] == 1
+            ), "concurrency for devtrain data should be 1"
+
+        for name, v in sorted(cv_data.items()):
+            self.add_corpus(name, data=v, add_lm=False)
+            self.cv_corpora.append(name)
+            assert self.concurrent[name] == 1, "concurrency for cv data should be 1"
+
+        for name, v in sorted(dev_data.items()):
+            self.add_corpus(name, data=v, add_lm=True)
+            self.dev_corpora.append(name)
+
+        for name, v in sorted(test_data.items()):
+            self.add_corpus(name, data=v, add_lm=True)
+            self.test_corpora.append(name)
+
+    # -------------------- Training --------------------
+
+    def nn_rasr_training(
+        self,
+        name: str,
+        train_corpus: str,
+        dev_corpus: str,
+        devtrain_corpus: str,
+        feature_corpus: str,
+        feature_flow: str,
+        alignment: str,
+        num_classes: int,
+        horovod_num_processes: int,
+        returnn_config: returnn.ReturnnConfig,
+    ):
+
+        self.train_nn(
+            name=name,
+            feature_corpus=feature_corpus,
+            train_corpus=train_corpus,
+            dev_corpus=dev_corpus,
+            feature_flow=feature_flow,
+            alignment=alignment,
+            num_classes=num_classes,
+            returnn_config=returnn_config,
+        )
+
+    # -------------------- Count LM Recognition --------------------
```

Comment:
Maybe we don't  define separate steps for "count recognition", "LSTM recognition" etc. but allow for an arbitrary number of recognitions with different settings.

Different settings (as specified by a dict, or data object) can be other type of LM, can be other prior (or -scale), ...

Reply:
Yes, I like that